packages/react-dom/src/__tests__/ReactDOMFragmentRefs-test.js

```diff
@@ -0,0 +1,557 @@
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @emails reactcore
+ */
+
+'use strict';
+
+let React;
+let ReactDOMClient;
+let act;
+let container;
+let Fragment;
+let Activity;
+
+describe('FragmentRefs', () => {
+  beforeEach(() => {
+    jest.resetModules();
+    React = require('react');
+    Fragment = React.Fragment;
+    Activity = React.unstable_Activity;
+    ReactDOMClient = require('react-dom/client');
+    act = require('internal-test-utils').act;
+    container = document.createElement('div');
+    document.body.appendChild(container);
+  });
+
+  afterEach(() => {
+    document.body.removeChild(container);
+  });
+
+  // @gate enableFragmentRefs
+  it('attaches a ref to Fragment', async () => {
+    const fragmentRef = React.createRef();
+    const root = ReactDOMClient.createRoot(container);
+
+    await act(() =>
+      root.render(
+        <div id="parent">
+          <Fragment ref={fragmentRef}>
+            <div id="child">Hi</div>
+          </Fragment>
+        </div>,
+      ),
+    );
+    expect(container.innerHTML).toEqual(
+      '<div id="parent"><div id="child">Hi</div></div>',
+    );
+
+    expect(fragmentRef.current).not.toBe(null);
+  });
+
+  // @gate enableFragmentRefs
+  it('accepts a ref callback', async () => {
+    let fragmentRef;
+    const root = ReactDOMClient.createRoot(container);
+
+    await act(() => {
+      root.render(
+        <Fragment ref={ref => (fragmentRef = ref)}>
+          <div id="child">Hi</div>
+        </Fragment>,
+      );
+    });
+
+    expect(fragmentRef.current).not.toEqual(null);
+  });
+
+  // @gate enableFragmentRefs
+  it('is available in effects', async () => {
+    function Test() {
+      const fragmentRef = React.useRef(null);
+      React.useLayoutEffect(() => {
+        expect(fragmentRef.current).not.toBe(null);
+      });
+      React.useEffect(() => {
+        expect(fragmentRef.current).not.toBe(null);
+      });
+      return (
+        <Fragment ref={fragmentRef}>
+          <div />
+        </Fragment>
+      );
+    }
+
+    const root = ReactDOMClient.createRoot(container);
+    await act(() => root.render(<Test />));
+  });
+
+  describe('focus()', () => {
+    // @gate enableFragmentRefs
+    it('focuses the first focusable child', async () => {
+      const parentRef = React.createRef();
+      const fragmentRef = React.createRef();
+      const root = ReactDOMClient.createRoot(container);
+      let focusedElement = null;
+
+      function Test() {
+        return (
+          <div ref={parentRef}>
+            <Fragment ref={fragmentRef}>
+              <div id="child-a" />
+              <a id="child-b" href="/">
+                B
+              </a>
+              <a id="child-c" href="/">
+                C
+              </a>
+            </Fragment>
+          </div>
+        );
+      }
+
+      await act(() => {
+        root.render(<Test />);
+      });
+
+      // The test environment doesn't implement focus.
```

Comment:
That sounds like a bug. JSDOM does implement focus. I removed your mock locally  but then focus ends up on the last child. `expect(document.activeElement.id).toEqual('child-b');` errors with
```
Expected: "child-b"
Received: "child-c"
```

`child-c` is just first child we try to focus when calling `fragmentInstance.focus()` because `getFragmentInstanceChildren` returns the children in reverse order.

The test currently only works because your focus mock doesn't dispatch a focus event so `setFocusIfFocusable` is always false.

Reply:
Thanks for catching this! I was using the wrong document methods to check the focused element and assumed the issue was in the environment. I switched the order of the traversal and removed the mocks in favor of asserting on `document.activeElement.id`